src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
+			if (state == 0)
+				StopPlayback();
+			else if (state == 1)
+				PrepareForPlayback();
+			else if (active_voices)
+				BeginPlayback();
 		}
+		CheckIrq();
+		return;
+	}
+
+	// All the registers below here involve voices
+	if (!voice)
+		return;
 
-		memset(&myGUS,0,sizeof(myGUS));
-		memset(GUSRam,0,1024*1024);
+	uint8_t data;
+	// Registers that write to the current voice
+	switch (selected_register) {
+	case 0x0: // Voice wave control register
+		if (voice->UpdateWaveState(register_data >> 8))
+			CheckVoiceIrq();
+		break;
+	case 0x1: // Voice rate control register
+		voice->WriteWaveRate(register_data);
+		break;
+	case 0x2: // Voice MSW start address register
+		UpdateWaveMsw(voice->wave_ctrl.start);
+		break;
+	case 0x3: // Voice LSW start address register
+		UpdateWaveLsw(voice->wave_ctrl.start);
+		break;
+	case 0x4: // Voice MSW end address register
+		UpdateWaveMsw(voice->wave_ctrl.end);
+		break;
+	case 0x5: // Voice LSW end address register
+		UpdateWaveLsw(voice->wave_ctrl.end);
+		break;
+	case 0x6: // Voice volume rate register
+		voice->WriteVolRate(register_data >> 8);
+		break;
+	case 0x7: // Voice volume start register  EEEEMMMM
+		data = register_data >> 8;
+		// Don't need to bounds-check the value because it's implied:
+		// 'data' is a uint8, so is 255 at most. 255 << 4 = 4080, which
+		// falls within-bounds of the 4096-long vol_scalars array.
+		voice->vol_ctrl.start = (data << 4) * VOLUME_INC_SCALAR;
+		break;
+	case 0x8: // Voice volume end register  EEEEMMMM
+		data = register_data >> 8;
+		// Same as above regarding bound-checking.
+		voice->vol_ctrl.end = (data << 4) * VOLUME_INC_SCALAR;
+		break;
+	case 0x9: // Voice current volume register
+		// Don't need to bounds-check the value because it's implied:
+		// reg data is a uint16, and 65535 >> 4 takes it down to 4095,
+		// which is the last element in the 4096-long vol_scalars array.
+		voice->vol_ctrl.pos = (register_data >> 4) * VOLUME_INC_SCALAR;
+		break;
+	case 0xA: // Voice MSW current address register
+		UpdateWaveMsw(voice->wave_ctrl.pos);
+		break;
+	case 0xB: // Voice LSW current address register
+		UpdateWaveLsw(voice->wave_ctrl.pos);
+		break;
+	case 0xC: // Voice pan pot register
+		voice->WritePanPot(register_data >> 8);
+		break;
+	case 0xD: // Voice volume control register
+		if (voice->UpdateVolState(register_data >> 8))
+			CheckVoiceIrq();
+		break;
 	}
-};
 
-static GUS* test;
+#if LOG_GUS
+	LOG_MSG("GUS: Unimplemented write register %x -- %x", register_select,
+	        register_data);
+#endif
+	return;
+}
 
-void GUS_ShutDown(Section* /*sec*/) {
-	delete test;	
+void GUS_ShutDown(Section * /*sec*/)
+{
+	if (gus) {
+		gus->PrintStats();
+		gus.reset(nullptr);
+	}
 }
 
-void GUS_Init(Section* sec) {
-	test = new GUS(sec);
-	sec->AddDestroyFunction(&GUS_ShutDown,true);
+void GUS_Init(Section *sec)
+{
+	assert(sec);
+	Section_prop *conf = dynamic_cast<Section_prop *>(sec);
+	if (!conf || !conf->Get_bool("gus"))
+		return;
```

Comment:
Just a suggestion: @kcgen what do you think about using the same pattern for configuration section, that I used for FluidSynth? (see `FLUID_AddConfigSection`). This way GUS-specific configuration could be defined in `gus.cpp` instead forcing reader to jump to `dosbox.cpp`.

Reply:
Will check out it!

Reply:
Much more compartmentalized! (and a good model for other devices, if/when they're worked-on).